lib/srv/db/common/permissions/import_rules.go

```diff
@@ -61,39 +65,121 @@ func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObject
-			// the matching is applied to the object spec; existing object labels does not matter
-			if !databaseObjectScopeMatch(mapping.GetScope(), obj.GetSpec()) {
-				continue
-			}
-			if databaseObjectImportMatch(mapping.GetMatch(), obj.GetSpec()) {
-				if objClone == nil {
-					objClone = utils.CloneProtoMsg(obj)
-				}
-
-				// mapping applies additional labels
-				labels := objClone.Metadata.Labels
-				if labels == nil {
-					labels = map[string]string{}
-				}
-				for k, v := range mapping.AddLabels {
-					labels[k] = v
-				}
-				objClone.Metadata.Labels = labels
+			if applyMappingToObject(mapping, objClone.GetSpec(), objClone.Metadata.Labels) {
+				matched = true
 			}
 		}
 
-		if objClone != nil {
+		if matched {
 			out = append(out, objClone)
 		}
 	}
 
 	return out
 }
 
+// splitExpressionRE matches the pattern "{{...}}"
+var splitExpressionRE = regexp.MustCompile(`{{[^}]*}}`)
+
+// splitExpressions splits a complex template like `foo-{{bar}}-{{baz}}` into its constituent parts,
+// which can be evaluated one by one.
+func splitExpressions(input string) []string {
+	indexes := splitExpressionRE.FindAllStringIndex(input, -1)
+
+	substringStart := 0
+	parts := make([]string, 0)
+
+	// iterate over matches
+	for _, matchIndex := range indexes {
+		// non-interpolated part
+		parts = append(parts, input[substringStart:matchIndex[0]])
+
+		// interpolated part
+		parts = append(parts, input[matchIndex[0]:matchIndex[1]])
+
+		// move start index
+		substringStart = matchIndex[1]
+	}
+
+	// remaining Append the remaining part of the string after the last match
+	parts = append(parts, input[substringStart:])
+
+	return parts
+}
+
+func evalMultiTemplate(input string, values map[string]string) (string, error) {
+	traits := map[string][]string{}
+	for k, v := range values {
+		traits[k] = []string{v}
+	}
+
+	var sb strings.Builder
+
+	parts := splitExpressions(input)
+	for _, part := range parts {
+		// skip empty parts
+		if part == "" {
+			continue
+		}
+
+		out, err := services.ApplyValueTraits(part, traits)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		if len(out) != 1 {
+			// this is unlikely to happen.
+			return "", trace.BadParameter("unexpected length of value array: %v", len(out))
+		}
+		sb.WriteString(out[0])
+	}
+
+	return sb.String(), nil
+}
+
+func applyMappingToObject(mapping *dbobjectimportrulev1.DatabaseObjectImportRuleMapping, spec *dbobjectv1.DatabaseObjectSpec, labels map[string]string) bool {
+	// the matching is applied to the object spec; existing object labels does not matter
+	if !databaseObjectScopeMatch(mapping.GetScope(), spec) {
+		return false
+	}
+	if !databaseObjectImportMatch(mapping.GetMatch(), spec) {
+		return false
+	}
+
+	for key, value := range mapping.AddLabels {
+		// remove any potential old values for this key.
+		delete(labels, key)
+
+		// evaluate multi-trait template.
+		out, err := evalMultiTemplate(value, objectSpecToTraits(spec))
+		if err != nil {
+			continue
```

Comment:
I don't think it's a good idea to silently drop the error. Ideally, the error should be propagated to the caller. Or at least log the error here. And in case it errors, should we delete the original value? Do we need special handling on `NotFound`

Reply:
I think deleting the previous value is better approach, as you are not getting a mixture of labels across different rules. It is a tough case, however, and I do wonder if we should just log the error and refuse to import that particular database object. This would be somewhat safer option: we cannot ever grant access to an object that wasn't imported.

Reply:
> I think deleting the previous value is better approach, as you are not getting a mixture of labels across different rules. It is a tough case, however, and I do wonder if we should just log the error and refuse to import that particular database object. This would be somewhat safer option: we cannot ever grant access to an object that wasn't imported.

Ya it's tough. I think I would still prefer to delete the label and log an error, so in case there is a higher rule that writes the label again, it can heal itself. Yet another hacky alternative is put an error in the label so users can see that in TAG. 

Is it possible to validate the template parts when the import rule is created? If so, we are more confident that it won't fail here.

Reply:
Yes, this is good point, I'll add validation to import rule creation. All other matters aside, this is good UX.

We should still handle error cases, just in case, but given the validation above I think the right call is to simply reject the object entirely, with appropriate log.

Reply:
Added. This got somewhat messy, as the dependencies went the wrong way, so I ended up moving a lot of code around. I believe it makes more sense now, along with fixing the accidental cyclic dependencies.